### DISCUSSION TITLE
Changement (inversion) des couleurs dans un menu

### DISCUSSION THREAD

**Discussion by Guillaume31a** [witsa/synapps#586] [Q&A] [answered]
Bonjour, je n'arrive pas à dissocier deux acteurs qui obéissent à la même règle. Je souhaiterais inverser les couleurs lorsque l'utilisateur passe sur un bouton et également lorsqu'il clique sur celui-ci.
Voici une capture de l'image du menu :

![image](https://user-images.githubusercontent.com/131860668/234522829-964af4a4-308b-48d9-b4e2-abc90c5788b9.png)

Lorsqu'on clique sur "documents" par exemple, nous avons cette fenêtre qui s'affiche à droite de l'écran :

![image](https://user-images.githubusercontent.com/131860668/234522602-9ebf7a95-8f24-484d-bbe0-b4d14a3cbbf9.png)

Petit soucis, le bouton cliqué, "document" **reste en blanc** alors que dans le menu ci dessus, on peut voir que les couleurs s'inversent lorsque je passe ou que je clique sur une des options (schéma elec, schéma hydro....) 

Voici une image de l'ensemble : 

![image](https://user-images.githubusercontent.com/131860668/234526153-36fa87c4-9a6c-4e01-bef1-2312b58c73cd.png)

et le code CSS correspondant au **thème de couleur** : 

![image](https://user-images.githubusercontent.com/131860668/234526358-5f3c4784-0b7a-4242-82c0-064e7504dc46.png)

J'ai essayé un nouveau code en CSS : 

![image](https://user-images.githubusercontent.com/131860668/234527174-a799e8c4-dd36-457b-a5df-3b7a4c9869d6.png)

mais lui inverse le problème, c'est à dire que les options (schéma elec, schéma hydro....) restent bleue lorsqu'elles sont cliquées mais le bouton "document" apparait bien lorsque l'utilisateur clique dessus (bleu) (voir screen en dessous, ma souris étant la flèche noire dessinée) : 

![image](https://user-images.githubusercontent.com/131860668/234529041-94d1c0ff-6455-4370-aa0d-0dc7c3c17497.png)

et pour les options : 

![image](https://user-images.githubusercontent.com/131860668/234529636-2ba0a334-8c92-4832-9cdd-42a4596b8508.png)

Sur cette image, je passe juste ma souris sur "analyse fonctionnelle" et les couleurs sont inversées. Tout vas bien ? Et non ! J'ai eu un espoir aussi mais j'ai rencontré un autre problème. Lorsque je clique sur "analyse fonctionnelle" le bouton redeviens bleu comme si l'utilisateur n'avais pas cliqué dessus et quand on repasse avec la souris sur ce même bouton, il reste bleu : 

![image](https://user-images.githubusercontent.com/131860668/234530840-c5ea6160-a7f3-414c-b423-ed2f49403f59.png)

Sur ce screen j'ai cliqué avec ma souri. le bouton est donc censé se mettre en blanc. (ce n'est pas le cas).

Mais lorsque je clique hors de l'interface menu, le bouton "analyse fonctionnelle" se met "à jour" ou je ne sais pas trop mais il s'actualise et se met en mode inversé, donc cliqué comme ceci : 

![image](https://user-images.githubusercontent.com/131860668/234533461-cfc0b355-9494-4672-9ab8-c63dc9031595.png)

**En somme, il faudrait que lorsque que je clique ou que je passe sur un bouton, sa couleur s'inverse sans avoir a cliquer ailleurs.**

J'espère avoir été assez clair dans mes explications, j'espère qu'une solution sera trouvée, merci !

**Comment by Raphwit** [comment]
Bonjour, je remarque quelques petites erreurs dans le code CSS.

Si je cerne bien votre problème, vous devriez avoir une seule et même classe qui gère le comportement sur vos boutons.

Nous aurions donc la classe 'myClass' par exemple qui indiquerai à l'élément de changer de background et de couleur de texte.
Un exemple ci-dessous : 
![image](https://user-images.githubusercontent.com/93767525/234546754-3b9f9eed-7319-4203-b2bb-a0746866faab.png)

Voici le code de l'acteur html à copier / coller dans votre scène : 
Il est à noter que vous êtes libre de changer le nom de la classe, ainsi que les différentes couleurs.
```
SYNAPPS-STUDIO-ACTOR|{"type":"display/html","key":"html1","properties":{"content":"<style>\n\n  .myClass:hover {\n    background-color: blue;\n    color: white;\n  }\n  \n</style>","cursor":"pointer"}}
```

Cette classe a donc pour effet de changer le background et la couleur de texte en blanc uniquement au survol de l'élément.

Nous ne devrions pas avoir à gérer le focus dans votre exemple selon moi.

Il faudra enfin ajouter la classe 'myClass' à votre acteur afin que celui-ci ai le comportement défini par cette classe.

**Reply by Guillaume31a** [reply]
Merci pour votre réponse, je viens de mettre le code dans cette rubrique : 
![image](https://user-images.githubusercontent.com/131860668/234571832-1f6344a0-8b81-496c-a1e1-9e32f20241aa.png)
Ce code marche lorsque je clique sur "document" ou "comptages" : 
![image](https://user-images.githubusercontent.com/131860668/234572308-23eb9789-8148-4a32-ae8f-035e3b5528d4.png)
![image](https://user-images.githubusercontent.com/131860668/234573075-c8e0660e-54e9-4bed-b759-0e4363ec5e59.png)
Mais lorsque je clique sur "réglages" par exemple, je rencontre le même problème et il ne s'affiche pas en bleu : 
![image](https://user-images.githubusercontent.com/131860668/234572940-a2ab6102-2c99-4a34-a510-aeba29d759b6.png)
Merci d'avance pour votre réponse

**Reply by Raphwit** [reply]
Le code que je vous ai fournit est un acteur 'HTML', il faut coller celui ci directement dans l'arborescence des acteurs dans synapps studio.

Par exemple comme-ceci :
![image](https://user-images.githubusercontent.com/93767525/234574825-eda45f2e-3887-46c7-a316-5301c7f90077.png)

**Reply by Guillaume31a** [reply]
j'ai oublié un détail, lorsque je clique sur Electricité CFO, il y a un menu déroulant avec réglage qui s'ouvre : 
avant clic : 
![image](https://user-images.githubusercontent.com/131860668/234574983-f6fbf5e1-1e4e-4978-863d-1ad667da9a0d.png)
apres clic : 
![image](https://user-images.githubusercontent.com/131860668/234575064-fed709fb-5389-48a2-b6c1-3db990684d7f.png)
je ne sais pas si ça influe sur le code.

**Reply by Raphwit** [reply]
Non, cela n'a pas d'influence, aucun soucis, il faut que vous colliez le code que j'ai envoyé plus haut et coller celui-ci dans l'arborescence des acteurs dans synapps studio.

Une fois ceci fait, vous devriez voir un nouvel acteur HTML contenant un certain style.

A partir de ce point, il vous suffit de rajouter la classe 'myClass' à vos boutons pour que ceux-ci ai le comportement souhaité

**Reply by Guillaume31a** [reply]
j'ai bien collé dans l'arborescence comme indiqué par la flèche blanche :  
![image](https://user-images.githubusercontent.com/131860668/234578701-b2f861ab-14b8-426e-a320-785974600725.png)
Pour ce qui est de la classe à changer, est-ce bien à l'endroit du point d'interrogation ? ça donnerait donc quelque chose comme : 
![image](https://user-images.githubusercontent.com/131860668/234579297-f6459e4c-83db-4ca9-90c3-904955e1375d.png)
Pour ce qui est du code CSS dans "composites" -> "thème de couleur", j'ai mis votre ligne de code et ça a l'air de mieux fonctionner : 
![image](https://user-images.githubusercontent.com/131860668/234579934-1d5ef83d-e28b-4666-aec8-a33a3f4ded92.png)
est-ce important de garder l'ancien CSS ou ce n'est absolument pas nécessaire ?

**Reply by Raphwit** [reply]
Sur le premier screen que vous m'avez envoyé, vous ne regardez pas le bon acteur, vous êtes sur le 'bouton de navigation' et non sur l'acteur 'HTML' que l'on vient de créer.

En ce qui concerne le code CSS dans "composite" -> "thème de couleur" il faut en réalité remplacer l'acteur HTML que vous avez par celui que je vous ai fourni, l'ancien n'est donc pas utile.

**Reply by Guillaume31a** [reply]
Dans votre message : 

**_A partir de ce point, il vous suffit de rajouter la classe 'myClass' à vos boutons pour que ceux-ci ai le comportement souhaité_**

ou faut il que je me rende ? 

Merci

**Reply by Raphwit** [reply]
Il faut renseigner le champ 'Classes' sous la catégorie 'Aspect'

![image](https://user-images.githubusercontent.com/93767525/234593869-01cb29fc-63cc-4983-ad23-2bb1317b5856.png)

**Reply by Guillaume31a** [reply]
Retour après tests : 
j'ai bien attribué la classe à 1 seul bouton "documents"
![image](https://user-images.githubusercontent.com/131860668/234603587-36f3ac27-8c98-4b9f-b828-41c4335f0746.png)
j'ai modifié les couleurs du code html de la sorte pour que ca soit bien voyant : 
![image](https://user-images.githubusercontent.com/131860668/234604019-61ade40f-6183-4385-b564-38c5158b1855.png)
Lorsque j'exécute le synapps dans le navigateur, je ne constate aucun changement de couleur (même quand je ne fait que passer la souris dessus) : 
![image](https://user-images.githubusercontent.com/131860668/234604481-5fe942fe-7af3-4c80-9af4-ac3f4dee1bc2.png)
J'ai donc essayé d'assigner la classe au stack qui prend en compte tout le bandeau de droite : 
![image](https://user-images.githubusercontent.com/131860668/234605626-2fa92293-42ee-4c62-9c3e-21dc125659b7.png)
et cette fois-ci j'ai ce rendu : 
![image](https://user-images.githubusercontent.com/131860668/234605125-ee8381be-7ca8-45a4-a672-484dacc527c8.png)
suis-je passé à coté de quelque chose  ?
Merci

**Reply by Raphwit** [reply, marked as answer]
Mes excuses, je n'ai pas été assez précis pour le code CSS, il faut remplacer le code à l'intérieur de l'acteur par celui-ci :

 ```
<style>

  .synapps-ui .actor button.myClass:hover {
    background-color: blue !important;
    color: white !important;
  }
</style>
```

Cela devrait fonctionner !

Voila un acteur HTML qui permet de changer le background et la couleur de texte au survol : 

```
SYNAPPS-STUDIO-ACTOR|{"type":"display/html","key":"html1","properties":{"content":"<style>\n\n  .synapps-ui .actor button.{{myClass}}:hover {\n    background-color: {{bg_color}} !important;\n    color: {{text_color}} !important;\n  }\n</style>","cursor":"pointer"},"additionalDefs":{"myClass":{"type":"text"},"bg_color":{"type":"color"},"text_color":{"type":"color"}},"additionals":{"myClass":"myClass","bg_color":"rgba(42, 0, 255, 1)","text_color":"rgba(255, 255, 255, 1)"}}
```

'myClass' est le nom de la classe à renseignée dans le champ dont nous parlions plus haut,
'bg_color' représente la couleur du background,
'text_color' la couleur du texte.

**Reply by Guillaume31a** [reply]
je teste ça demain matin et je reviens vers vous, merci beaucoup, très bonne fin de journée !

**Reply by Guillaume31a** [reply]
Bonjour, j'ai testé ce que vous m'avez envoyé et cela fonctionne ! 
J'avais une dernière question sur le même projet, j'aimerais que lorsque je clique sur un de mes acteurs (accueil, journal,courbes,cvc ....) il se mette validé en bleu et non en blanc/gris, merci !
![image](https://user-images.githubusercontent.com/131860668/234784066-af640029-3b45-4994-acf9-d24f1ad601e9.png)

**Reply by Raphwit** [reply]
Vous pouvez rajouter ce code au sein de l'acteur HTML que nous avons créée : 

```
.synapps-ui .actor button.current {
   background-color: blue !important;
   color: white !important;
 }
```

Avec toujours le même principe, la couleur reste entièrement personnalisable !

**Reply by Guillaume31a** [reply]
parfait merci beaucoup pour votre aide et votre patience !
très bonne journée.
